p2p/src/codec.rs

```diff
@@ -0,0 +1,284 @@
+                        ))
+                    }
+                }
+                self.state = DecodeState::Len(command_type);
+                self.decode(src)
+            }
+            DecodeState::Len(m_type) => {
+                if src.len() < 4 {
+                    return Ok(None);
+                }
+                let len = src.get_u32() as usize;
+                self.state = DecodeState::Body(m_type, len);
+                self.decode(src)
+            }
+            DecodeState::Body(m_type, len) => {
+                if src.len() < len {
+                    return Ok(None);
+                }
+                self.state = DecodeState::MessageType;
+                read_message_body(m_type, len, src).map(Some)
+            }
+        }
+    }
+}
+
+fn read_message_body(
+    message_type: u8,
+    len: usize,
+    src: &mut BytesMut,
+) -> Result<PeerMessage, io::Error> {
+    match message_type {
+        0 => {
+            if len != 2 {
+                return Err(io::Error::new(
+                    io::ErrorKind::InvalidData,
+                    format!("Invalid length for hello message: {}", len),
+                ));
+            }
+            let data = src.get_u16();
+            Ok(PeerMessage::Hello(data))
+        }
+        1 => {
+            let mut peers = vec![];
+            let mut peers_bytes = src.split_to(len);
+            while !peers_bytes.is_empty() {
+                let peer = decode_peer_addr(&mut peers_bytes)?;
+                peers.push(peer);
+            }
+            Ok(PeerMessage::Peers(peers))
+        }
+        2 => Ok(PeerMessage::Data(src.split_to(len).to_vec())),
+        m => Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            format!("Unknown message type: {}", m),
+        )),
+    }
+}
+
+fn encode_peer_addr(peer: PeerAddr, buf: &mut BytesMut) {
+    match peer.addr {
+        SocketAddr::V4(d) => {
+            buf.put_u8(4);
+            buf.put(&d.ip().octets()[..]);
+            buf.put_u16(d.port());
+        }
+        SocketAddr::V6(d) => {
+            buf.put_u8(6);
+            buf.put(&d.ip().octets()[..]);
+            buf.put_u16(d.port());
+            buf.put_u32(d.flowinfo());
+            buf.put_u32(d.scope_id())
+        }
+    }
+    buf.put(peer.id.0.as_bytes());
+}
+
+fn decode_peer_addr(buf: &mut BytesMut) -> Result<PeerAddr, io::Error> {
+    let addr = read_socket_addr(buf)?;
+    if buf.len() < 32 {
+        return Err(io::Error::new(
+            io::ErrorKind::InvalidData,
+            format!("Must be 32 bytes for key, but found {}", buf.len()),
+        ));
+    }
+    let key = buf.split_to(32);
+    let id = PeerID(PublicKey::from(CompressedRistretto::from_slice(
+        key.as_ref(),
+    )));
+    Ok(PeerAddr { id, addr })
+}
+
+fn read_socket_addr(buf: &mut BytesMut) -> Result<SocketAddr, io::Error> {
+    // We check if !buf.is_empty() in MessageDecoder::decode, so it have at least 1 byte
+    let ipv = buf.get_u8();
```

Comment:
I'd keep functions as robust locally as possible, even if we do overlapping checks - those are not expensive, but later allowing a crash may be devastating. Let's add a check for enough bytes.

To make all these checks not so mouthful, i'd introduce `fn check_length(&mut BytesMut, n: usize, label: &str) -> Result<(), io::Error>` that simply errs with UnexpectedEof and message "need {n} bytes for {label}, but only {buf.len} is available in the buffer". Then we can do one-liners like `check_length(buf, 1)?;` before any attempt to read.

Reply:
Done.